crucible-llvm/src/Lang/Crucible/LLVM/SymIO.hs

```diff
@@ -0,0 +1,560 @@
+import qualified Data.Parameterized.NatRepr as PN
+import qualified Data.Parameterized.SymbolRepr as PS
+import qualified Data.Set as Set
+import qualified Data.Text as Text
+import qualified Data.Text.Encoding as TextEncoding
+import qualified Data.Text.IO as Text.IO
+import           GHC.Natural ( Natural )
+import           GHC.TypeNats ( type (<=) )
+import qualified System.IO as IO
+
+import qualified Lang.Crucible.FunctionHandle as LCF
+import           Lang.Crucible.Types ( IntrinsicType, BVType, TypeRepr(..) )
+import qualified Lang.Crucible.Utils.MuxTree as CMT
+import           Lang.Crucible.CFG.Common
+import           Lang.Crucible.Backend as C
+import           Lang.Crucible.Simulator.OverrideSim
+import           Lang.Crucible.Simulator.Intrinsics
+import           Lang.Crucible.Simulator.RegMap
+import qualified Lang.Crucible.Simulator.GlobalState as LCSG
+
+import           Lang.Crucible.LLVM.Bytes (toBytes)
+import           Lang.Crucible.LLVM.MemModel
+import           Lang.Crucible.LLVM.Extension ( ArchWidth )
+import           Lang.Crucible.LLVM.DataLayout ( noAlignment )
+import           Lang.Crucible.LLVM.Intrinsics
+import           Lang.Crucible.LLVM.QQ( llvmOvr )
+
+import qualified What4.Interface as W4
+import qualified What4.Partial as W4P
+
+import qualified Lang.Crucible.SymIO as SymIO
+
+-- | A representation of the filesystem for the LLVM frontend
+--
+-- This contains the underlying SymIO filesystem as well as the infrastructure
+-- for allocating fresh file handles
+data LLVMFileSystem ptrW =
+  LLVMFileSystem
+    { llvmFileSystem :: GlobalVar (SymIO.FileSystemType ptrW)
+    -- ^ The underlying symbolic filesystem
+    , llvmFileDescMap :: GlobalVar (FDescMapType ptrW)
+    -- ^ Maintains the mapping from file descriptors to low-level crucible-symio
+    -- 'SymIO.FileHandle's
+    , llvmHandles :: Map.Map Natural IO.Handle
+    -- ^ Handles that concrete output will be mirrored to; if this is empty, no
+    -- mirroring will be performed
+    , llvmFilePointerRepr :: PN.NatRepr ptrW
+    }
+
+-- | Contains the mapping from file descriptors to the underlying 'SymIO.FileHandle'
+--
+-- This also tracks the next file descriptor to hand out.  See Note [File
+-- Descriptor Sequence].
+data FDescMap sym ptrW where
+  FDescMap ::
+    { fDescNext :: Natural
+    -- ^ The next file descriptor to hand out
+    --
+    -- Note that these are truncated to 32 bit bitvectors when they are handed
+    -- out; we don't have any guards against overflow on that right now
+    , fDescMap :: Map.Map Natural (W4P.PartExpr (W4.Pred sym) (SymIO.FileHandle sym ptrW))
+    } -> FDescMap sym ptrW
+
+-- | A wrapper around a RankN 'OverrideSim' action
+--
+-- This enables us to make explicit that all of the type variables are free and
+-- can be instantiated at any types since this override action has to be
+-- returned from the 'initialLLVMFileSystem' function.
+newtype SomeOverrideSim sym a where
```

Comment:
This might be a useful utility definition, and better placed elsewhere...

Reply:
Opened #801